tensorflow/core/kernels/tensor_forest/resources.cc

```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
+  }
+  return false;
+};
+
+void TensorForestFertileStatsResource::UpdateSlotStats(
+    const bool is_regression, const int32 node_id, const int32 example_id,
+    const int32 num_targets, const TTypes<float>::ConstMatrix* dense_feature,
+    const TTypes<float>::ConstMatrix* labels) {
+  float incoming_weight = 1.0;
+  auto slot = fertile_stats_->node_to_slot().at(node_id);
+
+  auto label = (*labels)(example_id, 0);
+  float old_total_weight = slot.leaf_stats().counts_or_sums().value(label);
+
+  slot.mutable_leaf_stats()->set_weight_sum(incoming_weight +
+                                            slot.leaf_stats().weight_sum());
+  slot.mutable_leaf_stats()->mutable_counts_or_sums()->set_value(
+      label, old_total_weight + incoming_weight);
+
+  for (auto candidate : slot.candidates()) {
+    float old_left_weight =
+        candidate.left_leaf_stats().counts_or_sums().value(label);
+
+    if ((*dense_feature)(example_id, candidate.split().feature_id()) >=
+        candidate.split().threshold()) {
+      candidate.mutable_left_split_stats()->mutable_sum()->set_value(
+          label,
+          candidate.left_split_stats().sum().value(label) + incoming_weight);
+
+      float new_weight = old_left_weight + incoming_weight;
+
+      candidate.mutable_left_split_stats()->mutable_sum_of_square()->set_value(
+          0, candidate.left_split_stats().sum_of_square().value(0) -
+                 old_left_weight * old_left_weight + new_weight * new_weight);
+
+      candidate.mutable_left_leaf_stats()->set_weight_sum(
+          incoming_weight + candidate.left_leaf_stats().weight_sum());
+
+      candidate.mutable_left_leaf_stats()->mutable_counts_or_sums()->set_value(
+          label, new_weight);
+
+    } else {
+      float old_right_weight = old_total_weight - old_left_weight;
+
+      candidate.mutable_right_split_stats()->mutable_sum()->set_value(
+          label,
+          candidate.right_split_stats().sum().value(label) + incoming_weight);
+
+      float new_weight = old_right_weight + incoming_weight;
+
+      candidate.mutable_right_split_stats()->mutable_sum_of_square()->set_value(
+          0, candidate.right_split_stats().sum_of_square().value(0) -
+                 old_right_weight * old_right_weight + new_weight * new_weight);
+    }
+  }
+};
+
+const bool TensorForestFertileStatsResource::BestSplitFromSlot(
+    const tensor_forest::FertileSlot& slot,
+    tensor_forest::SplitCandidate* best) const {
+  float min_score = FLT_MAX;
+  for (int i = 0; i < slot.candidates_size(); i++) {
+    auto candidate = slot->candidates(i);
+    // for (auto candidate : slot.candidates()) {
+    float left_gini =
+        1 - candidate.left_split_stats().sum_of_square().value(0) /
+                (candidate.left_split_stats().sum().value(0) *
+                 candidate.left_split_stats().sum().value(0));
+    float right_gini =
+        1 - candidate.right_split_stats().sum_of_square().value(0) /
```

Comment:
can u make getGini a method that takes in split stats? This way it is much shorter and much less error prone